runtime/gc_realtime/Scheduler.cpp

```diff
@@ -762,6 +762,19 @@ void MM_Scheduler::yieldFromGC(MM_EnvironmentRealtime *env, bool distanceChecked
 	}
 }
 
+void
+MM_Scheduler::run(MM_EnvironmentBase *env, MM_Task *task, uintptr_t newThreadCount)
+{
+	uintptr_t activeThreads = recomputeActiveThreadCountForTask(env, task, newThreadCount);
+	task->mainSetup(env);
+	prepareThreadsForTask(env, task, activeThreads);
+	acceptTask(env);
+	task->run(env);
+	completeTask(env);
+	cleanupAfterTask(env);
+	task->mainCleanup(env);
+}
+
 void
 MM_Scheduler::prepareThreadsForTask(MM_EnvironmentBase *env, MM_Task *task, uintptr_t threadCount)
```

Comment:
It would be nice if this prepareThreadsForTask was just calling
- base Dispatcher prepareThreadsForTask
- pushYieldCollaborator
Then all these extra changes would not be required. But I don't understand why there are other differences  between this and base prepareThreadsForTask, and it's not worthwhile investing time in understanding that.

Also, I suspect just simply moving recomputeActiveThreadCountForTask and mainSetup into Scheduler's prepareThreadsForTask first could work (hence doing recomputeActiveThreadCountForTask twice (mainSetup does nothing of RT tasks) for a short while), but ok, let's play safe....

Reply:
>Also, I suspect just simply moving recomputeActiveThreadCountForTask and mainSetup into Scheduler's prepareThreadsForTask first could work (hence doing recomputeActiveThreadCountForTask twice (mainSetup does nothing of RT tasks) for a short while)

I considered this, but it seems like moving `recomputeActiveThreadCountForTask` inside `prepareThreadsForTask` could be problematic because recomputeActiveThreadCountForTask would be done twice (as you mentioned)... but the 2nd call will not take correct input, the 2nd call would take input returned by the first call (passed from run method).... so essentially, each call will have different input and hence expecting different output.